presto-delta/src/main/java/com/facebook/presto/delta/DeltaMetadata.java

```diff
@@ -0,0 +1,275 @@
+    /**
+     * Special schema used when querying a Delta table by storage location.
+     * Ex. SELECT * FROM delta."$PATH$"."s3://bucket/path/to/table". User is not able to list any tables
+     * in this schema. It is just used to query a Delta table by storage location.
+     */
+    private static final String PATH_SCHEMA = "$PATH$";
+
+    private final String connectorId;
+    private final DeltaClient deltaClient;
+    private final ExtendedHiveMetastore metastore;
+    private final TypeManager typeManager;
+    private final DeltaConfig config;
+
+    @Inject
+    public DeltaMetadata(
+            DeltaConnectorId connectorId,
+            DeltaClient deltaClient,
+            ExtendedHiveMetastore metastore,
+            TypeManager typeManager,
+            DeltaConfig config)
+    {
+        this.connectorId = requireNonNull(connectorId, "connectorId is null").toString();
+        this.deltaClient = requireNonNull(deltaClient, "deltaClient is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.config = requireNonNull(config, "config is null");
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        ArrayList<String> schemas = new ArrayList<>();
+        schemas.addAll(metastore.getAllDatabases(metastoreContext(session)));
+        schemas.add(PATH_SCHEMA.toLowerCase(Locale.ROOT));
+        return schemas;
+    }
+
+    @Override
+    public DeltaTableHandle getTableHandle(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        String schemaName = schemaTableName.getSchemaName();
+        String tableName = schemaTableName.getTableName();
+        if (!listSchemaNames(session).contains(schemaName)) {
+            return null; // indicates table doesn't exist
+        }
+
+        DeltaTableName deltaTableName = DeltaTableName.from(tableName);
+        String tableLocation;
+        if (PATH_SCHEMA.equalsIgnoreCase(schemaName)) {
+            tableLocation = deltaTableName.getTableNameOrPath();
+        }
+        else {
+            Optional<Table> metastoreTable = metastore.getTable(metastoreContext(session), schemaName, deltaTableName.getTableNameOrPath());
+            if (!metastoreTable.isPresent()) {
+                return null; // indicates table doesn't exist
+            }
+
+            Map<String, String> tableParameters = metastoreTable.get().getParameters();
+            Storage storage = metastoreTable.get().getStorage();
+            tableLocation = storage.getLocation();
+
+            // Delta table written using Spark and Hive have set the table parameter
+            // "spark.sql.sources.provider = delta". If this property is found table
+            // location is found in SerDe properties with key "path".
+            if ("delta".equalsIgnoreCase(tableParameters.get("spark.sql.sources.provider"))) {
```

Comment:
shall we define `delta`, `spark.sql.sources.provider` as constants? or static import?

Reply:
These constants are used only here and also there is a comment that uses the same constant. I am thinking it is better to leave it as it is. Let me know.